### PR TITLE
#159054020 Build out articles pagination 

### PR DESCRIPTION
**Pull request by denniswanjiru**
#### What does this PR do?
Paginating articles

#### Description of Task to be completed?
- Paginate all the articles

#### How should this be manually tested?
- Clone this branch

```
git clone -b  ft-pagination-159054020 git@github.com:andela/ah-jarvis.git

cd ah-jarvis
  
create and activate a virtual env 

pip install -r requirements.txt

python manage.py migrate

python manage.py runserver

```
to test make an api con on this endpont `GET /api/articles/?limit=<int>&offset=<int>`
##### Articles response 
screen shot 2018-08-16 at 10 59 01

#### What are the relevant pivotal tracker stories?
[#159054020](https://www.pivotaltracker.com/story/show/159054017/comments/159054020)

### CODE REVIEW

**Comment by jacksono** [comment]
@DennisWanjiru please name the pagination query param `limit` instead of `page`.
Also if you look at the API specs on the repo  readme, there's an offset query param which you should implement.
Did you include the options for next and previous pages?

**Comment by denniswanjiru** [comment]
Okay, I will make those changes. And no, I didn't include the options for next and prev pages. Django does that when you specify explicitly the number of articles you want returned per page.

**Comment by jacksono** [comment]
@DennisWanjiru ok. I had not noticed the magic of Django. Then just work on `limit` and `offset` and let me know when its ready

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -9,12 +10,19 @@
...
 
 from .models import Article, Rate, Comment
 from .serializers import ArticleSerializer, CommentSerializer, RateSerializer
 from .renderers import ArticleJSONRenderer, CommentJSONRenderer, RateJSONRenderer, FavoriteJSONRenderer
 
 
+class LargeResultsSetPagination(PageNumberPagination):
+    page_size = os.getenv('PAGE_SIZE')
+    page_size_query_param = 'page_size'
+    max_page_size = os.getenv('PAGE_SIZE')
```
Comment (asamisellah) [inline authors/apps/articles/views.py]: Do we need to set these variables in our .env files?

**Comment by asamisellah** [comment]
I have tested the feature and it is not working as expected.

**Comment by denniswanjiru** [comment]
@asamisellah I just fixed it try now

**Comment by Duncanian** [comment]
I have just tested and it works well on my end.

**Comment by denniswanjiru** [comment]
@jacksono This PR is ready for final review, thanks.

**Comment by asamisellah** [comment]
@DennisWanjiru I've tested the feature after the changes you've made and it works as expected.

**Comment by Tittoh** [comment]
please include the parameters for the endpoint on your PR for example `GET /api/articles?title=<title>`

**Comment by Tittoh** [comment]
Tested it and it's working as expected.

**Comment by jacksono** [comment]
@Tittoh  what did you mean by your first comment?

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 """ Views for django Articles. """
+import os
```
Comment (jacksono) [inline authors/apps/articles/views.py]: Is there anywhere you are using the `os` you have imported?

Comment (denniswanjiru) [reply authors/apps/articles/views.py]: I was using in the previous version. I forgot to pluck it out. Thanks for the reminder.

**Comment by Tittoh** [comment]
@jacksono the arguments to pass on the url were missing on the PR. It has been added. `GET /api/articles/?limit=<int>&offset=<int>`

File: authors/settings.py [hunk 3]
```diff
@@ -162,8 +162,8 @@
     'DEFAULT_AUTHENTICATION_CLASSES': (
         'authors.apps.authentication.backends.JWTAuthentication',
     ),
-    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.LimitOffsetPagination'
-
+    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.LimitOffsetPagination',
+    'PAGE_SIZE': 9
```
Comment (jacksono) [inline authors/settings.py]: Lets wait and see if the PO has an opinion on this default limit. The rest is ok. Do you intend to add any unit tests for this?

Comment (denniswanjiru) [reply authors/settings.py]: Now that I have added the default limit, I think I need to test it.

File: authors/apps/articles/tests/test_pagination.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            "article": {
+                "title": "How to feed your dragon",
+                "description": "Wanna know how?",
+                "body": "You don't believe?",
+            }
+        }
+
+    def login_user(self, user=user):
+        """
+        login user
```
Comment (Duncanian) [inline authors/apps/articles/tests/test_pagination.py]: Use a capital to begin the docstring

**Review by Duncanian — CHANGES_REQUESTED** [review]

**Review by Duncanian — DISMISSED** [review]

**Comment by denniswanjiru** [comment]
@jacksono The PO just advised me to have 12 articles per page and I have made those change. I think this PR is therefore ready for final review. Thanks

**Review by jacksono — APPROVED** [review]

**Comment by dwarukira** [comment]
The Implementation looks good. But maybe you can make the pagination work on comments

**Pull request merged**